Documentation/ABI/TestContent.md

```diff
@@ -0,0 +1,178 @@
+
+See https://swift.org/LICENSE.txt for license information
+See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+-->
+
+This document describes the format and location of test content that the testing
+library emits at compile time and can discover at runtime.
+
+> [!WARNING]
+> The content of this document is subject to change pending efforts to define a
+> Swift-wide standard mechanism for runtime metadata emission and discovery.
+> Treat the information in this document as experimental.
+
+## Basic format
+
+Swift Testing stores test content records in a dedicated platform-specific
+section in built test products:
+
+| Platform | Binary Format | Section Name |
+|-|:-:|-|
+| macOS, iOS, watchOS, tvOS, visionOS | Mach-O | `__DATA_CONST,__swift5_tests` |
+| Linux, FreeBSD, Android | ELF | `swift5_tests` |
+| WASI | Statically Linked | `swift5_tests`[^1] |
+| Windows | PE/COFF | `.sw5test`[^2] |
+
+[^1]: SwiftWasm effectively uses the ELF format for its images, however it is
+      currently always statically linked and runtime discovery is performed
+      using a different mechanism than what we use on other ELF-based platforms.
+[^2]: On Windows, the Swift compiler [emits](https://github.com/swiftlang/swift/blob/main/stdlib/public/runtime/SwiftRT-COFF.cpp)
+      leading and trailing padding into this section, both zeroed and of size
+      `MemoryLayout<UInt>.stride`. Code that walks this section can safely skip
+      over this padding.
+
+### Record headers
+
+Regardless of platform, all test content records created and discoverable by the
+testing library have the following layout:
+
+```swift
+typealias TestContentRecord = (
+  kind: UInt32,
+  version: UInt16,
+  reserved1: UInt16,
+  accessor: (@convention(c) (_ outValue: UnsafeMutableRawPointer, _ hint: UnsafeRawPointer?) -> CBool)?,
+  context: UInt,
+  reserved2: UInt
+)
+```
+
+This type has natural size, stride, and alignment. Its fields are native-endian.
+If needed, this type can be represented in C as a structure:
+
+```c
+struct SWTTestContentRecord {
+  uint32_t kind;
+  uint16_t version;
+  uint16_t reserved1;
+  bool (* _Nullable accessor)(void *outValue, const void *_Null_unspecified hint);
+  uintptr_t context;
+  uintptr_t reserved2;
+};
+```
+
+### Record contents
+
+#### The kind field
+
+Each record's _kind_ determines how the record will be interpreted at runtime. A
+record's kind is a 32-bit unsigned value. The following kinds are defined:
+
+| As Hexadecimal | As [FourCC](https://en.wikipedia.org/wiki/FourCC) | Interpretation |
+|-:|:-:|-|
+| `0x00000000` | &ndash; | Reserved (**do not use**) |
+| `0x74657374` | `'test'` | Test or suite declaration |
+| `0x65786974` | `'exit'` | Exit test |
+
+<!-- When adding cases to this enumeration, be sure to also update the
+corresponding enumeration in TestContentGeneration.swift. -->
+
+#### The version field
+
+This field is currently always `0`. Implementations should ignore structures
```

Comment:
I think that it should be stronger and say that the field must be zero. But, I think that the question above also influences what this text should read. It could be the initial release would have different versions for the different kinds (unless this is per-record).

Reply:
Moot if I remove the field.